### ISSUE TITLE
An easy tool to improve translations

### ISSUE THREAD

**Issue by Meagain13** [codecombat/codecombat#6026]
Hi Team!

I wonder if it would be possible you to set a filter on one section of this link https://codecombat.com/i18n, since it would be much easier for me to translate as a Diplomat if the Type section could unfold the types (Level, Course, Achievement and so on), instead of scrolling down the whole list every time.

Just let me know if it´s possible or if there´s a quickest option I´m maybe missing.

Thanks a lot!

**Comment by pbodnar** [comment]
Oh, I've just tried the "translations page" at https://codecombat.com/i18n and I must say it's _horrible_ (see below). I think the translation efforts generally would need pretty more improvements than just "folding the types", moreover when there seems to be no introductory material for it. Otherwise only very patient translators will help, the others will quickly run away, IMHO.

So just to name a few ideas:

## Generally

1. Make the difference between GitHub translations and the translations page more evident: Now the page https://codecombat.com/contribute/diplomat states just: `You can start translating our levels by going to our translations page, or our interface and website on GitHub. Find your language locale file on GitHub, edit it online, and submit a pull request. `. If you don't read it carefully, you might be under a wrong impression that *alternative methods of translating the levels* are being listed.
2. Making at least a short tutorial on "how to approach and contribute translations" would be also helpful.

## Main view of the translations page

1. Add possibility to filter and search in the list, i. e. enable to search through the keys & translations (a possible inspiration [Crowdin](https://crowdin.com/)). Right now it is pretty hard to correct some text one finds in the game, because one needs to know / guess the exact name of the section to translate. Another reason is to have inspiration of how existing translations look like. Otherwise, without a dedicated "glossary" page for a given language, it is hard to maintain the translations consistent. The translations also don't seem to google-able.
2. Explain what the column "Specifically Covered" means. Or just rename it to "Fully translated" or similar if that is the real meaning of the column. Or better, put percentages of translated texts there together with a rename.

## Section translation view - Change(s) preview dialog

1. The "Expert View" doesn't handle characters with accents (diacritics) correctly. Instead of one UTF-8 character, it shows those messy "pseudo-characters".
1. It is only after reloading a given page when the differences are shown correctly both in the "Old vs. New" and "Expert view". Before that, the changes are presented quite wrongly, like showing that a character was doubled when it was just added in reality.

I know the above could be written down into multiple issues, so please take this as an initial wrap-up...

**Comment by Meagain13** [comment]
I totally agree with you pbodnar, and fully support all your suggestions, in fact I could invite great translators from all languages to come and translate if the conditions you mention were implemented, but I know they wouldn´t come if I told them now and they saw (as you said) how disencouraging the current method is. 

So let´s see if we can see those improvements applied soon, so that we can also improve the scope and quality of this already amazing game.

Meanwhile I´ll still be available in case someone wants to discuss more about improvements on this important issue, feel free to contact!

**Comment by pbodnar** [comment]
Thanks for your feedback, @Meagain13. I guess that nowadays nearly everyone already knows English, moreover when talking about those getting into IT, so the translators' environment won't be of a big priority for the game authors?

BTW another improvement idea: What about letting translators preview their changes in a given text _beforehand_, i. e. before submitting them?

And not that I would be a dedicated translator, but at least for testing it, I submitted 3 translations' corrections a week ago and I have no clue when they will be reviewed (and hopefully accepted). They are still in the "pending" state now...

**Comment by nwinter** [comment]
Thanks @Meagain13 and @pbodnar, this is great stuff! I've fixed a few of these, enumerated them all to organize the work:
- [ ] On /i18n, add a filter for translatable model types
- [x] On /contribute/diplomat, update the instructions
- [x] On the wiki, update the instructions to highlight the info from the i18n subpage
- [ ] On /i18n, make a way to find a specific text string to translate
- [x] On /i18n, explain what specifically/generally covered mean
- [x] On /i18n, show percentage completion for specifically/generally covered
- [ ] On patch diff preview modal, fix text encoding
- [ ] On /i18n model views, show a live preview of what the translation will look like in-game
- [x] On /i18n, suggest using direct.codecombat.com to bypass CDN caching issues after submission
- [ ] Catch up on /i18n translation patches
- [ ] Catch up on GitHub translation pull requests

**Comment by pbodnar** [comment]
@nwinter, thanks for the efforts started, looking forward to the results. 👍 

BTW I have just found issue #5816 that gives some similar ideas, a good source of inspiration, I think.

> On /i18n model views, show a live preview of what the translation will look like in-game

Well, that would be possibly great. Although by previewing changes, I actually meant "just" to be able to show the diff before one submits its translations on the translations page. :)

**Comment by Meagain13** [comment]
Good to know you are taking care of these issues, @nwinter 

Thanks a lot!

I´m also looking forward to the results, I bet they will help us translate much more easily.

**Comment by Meagain13** [comment]
Hi @nwinter 

Thanks a lot for your work, I´ve seen the changes you did on the /i18n web pages, on the wiki and on the diplomats instructions, so it´s easier and more comprenhensible now.

I´ve just also done some translations on the Github (es-ES.coffee file) and submited a pull request, but it´s my first time doing it so I don´t know if I did it right or wrong, could you let me know if it´s ok, so that I can go on translating?

Thanks in advance!

Cheers

**Comment by nwinter** [comment]
@Meagain13 thanks, we are trying to make some time to get caught up on these! Will comment on that PR.

**Comment by djanym** [comment]
Hey @nwinter 
Are there any expectations on this "On /i18n, make a way to find a specific text string to translate"?
The https://codecombat.com/i18n list became too long :)
Thanks.

**Comment by nwinter** [comment]
Not yet, sorry!
